### ISSUE TITLE
Support `@around` callbacks

### ISSUE THREAD

**Issue by schuetzm** [rejectedsoftware/vibe.d#1037]
These should work similar to `@before` and `@after`. They need to get passed a delegate that calls the actual method.

Use cases:
* catch exceptions and turn them into appropriate HTTP error codes
* locking
* temporarily change task-local state (e.g. timezone, i18n default language) and reset it using `scope(exit)` or `finally`

**Comment by Geod24** [comment]
Where would we put such attributes ? Could you give a (theoric) code example ?

**Comment by schuetzm** [comment]
```
interface ExampleAPI {
    @around!validateUTF8
    Bson test(string id, int x);
}

Ret validateUTF8(Ret, Args...)(
    HTTPServerRequest req, HTTPServerResponse res,
    scope Ret delegate(Args) yield, Args args
) {
    foreach(arg; args) {
        static if(is(typeof(arg) == string)) {
            enforceBadRequest(arg.isValidUTF8);
        }
    }
    try {
        return yield(args);
    } catch(UTFException) {
        throw new HTTPStatusException(400); // Bad Request
    }
}
```

That's how I imagine it at least. The callback has access to all the decoded parameters, just as the `test()` would receive them, and can adjust them as it sees fit. It can also manipulate whatever `test()` returns, in this case a `Bson` object.

It makes even more sense if it can be attached to several/all methods at once, to keep it DRY.

**Comment by s-ludwig** [comment]
UTF-8 validation should be integral part of the core system. Any more specific validation should be done using proper validation types like the one in `vibe.web.validation`.

Locking could theoretically be achieved using   a `synchronized` class, but IMO those are generally broken by design, so even if I'm not sure if this would be the best approach, I'd still consider this a semi-valid use case (semi because it can be achieved using other means, too, each with their own pros and cons).

> temporarily change task-local state (e.g. timezone, i18n default language) and reset it using scope(exit) or finally

Why would you do that? Is it common enough that a local solution within the interface function is not sufficient? To me personally it smells like more explicit code would not be a bad idea for such things, because due to all the hidden behavior it might easily be very confusing for someone reviewing such a code for the first time.

**Comment by schuetzm** [comment]
Agree about UTF8, but the more specific validations need to be handled in a very flexible manner. It simply isn't sufficient just to return an HTTP status code. As an example, I might want to add specific HTTP headers with information about what exactly didn't validate and why, that can be evaluated by the client.

Locking isn't restricted to mutexes; there's also file locks, MySQL table locks, and various others.

As for task-local state, I guess that's a question of philosophy. I agree that it can abused easily. But some things probably lend themselves more to it than others. Timezone and language are good examples IMO. They need to be set according to the preferences of a logged-in users, or maybe based on IP location guessing. And they will affect i18n string lookup in Diet templates, which means you would need to pass around a context object anyway. Thinking more about that though, `@before` should be sufficient for that, as long as a new request start a new (properly initialized) task the state doesn't need to be reset.

**Comment by Geod24** [comment]
@schuetzm : I've always been an opponent of `@before` / `@after`, at least on REST interfaces.

The problem with them is that it puts a server-only problem to the interface, and that breaks everything. Try to create a function that authenticate an user through MySQL for example, and use `@before` with it. You can't have the server and the client as separate applications, because since you're using the interface in both, you pull a reference to your function, which in turns pull a reference to every imported function. It is a serious design problem for me, and no attribute should be put on the interface if it's not part of the REST interface.

In your case, it seems your issues are server-side only, so those attributes should be put on the implementation. But I'll need a little more time to think about it :)

**Comment by s-ludwig** [comment]
> Agree about UTF8, but the more specific validations need to be handled in a very flexible manner. It simply isn't sufficient just to return an HTTP status code. As an example, I might want to add specific HTTP headers with information about what exactly didn't validate and why, that can be evaluated by the client.

`vibe.web.web` interface methods have access to the name and error messages for fields failing validation through the `@errorPage` hander (as well as access to an exception that was thrown). We should have something similar for REST, too. How the error is reported to the client can then be customized easily.

> Locking isn't restricted to mutexes; there's also file locks, MySQL table locks, and various others.

Right, `synchronized` classes were not part of my argument anyway, because I think that they are broken. Maybe I'm missing the proper use case, but most of the time I'd favor a RAII based solution (possibly in conjunction with `@before` instead of `@around`, and possibly with a monitor style locking mechanism to enforce locked access) or a local solution in the code (since you often only want parts of the code to be performed in locked state).

But as said, I don't consider this use case invalid or anything like that.

> As for task-local state, I guess that's a question of philosophy. I agree that it can abused easily. But some things probably lend themselves more to it than others. Timezone and language are good examples IMO. They need to be set according to the preferences of a logged-in users, or maybe based on IP location guessing. And they will affect i18n string lookup in Diet templates, which means you would need to pass around a context object anyway. Thinking more about that though, @before should be sufficient for that, as long as a new request start a new (properly initialized) task the state doesn't need to be reset.

Exactly, that was my main open question, why resetting the initial value would be needed. Since you'd usually always set the TZ/language setting for each request, even when no new task gets started for a request, it should still work with only `@before`, right?

**Comment by s-ludwig** [comment]
@Geod24 I agree, `@before`/`@after` can be great tools, but higher level annotations that give meaning to both client and server should be the way to go in most cases.

BTW, a useful thing might be to have something like `@after`, but for client requests. That way it would at least be possible to manually put both, server and client, behavior into the interface using annotations. Would be great if it were possible to combine multiple UDAs into one, though.

**Comment by Geod24** [comment]
@s-ludwig : See http://forum.rejectedsoftware.com/groups/rejectedsoftware.vibed/thread/24595/#post-24606
